File: 2021_summer_algorithm_study_intermediate/using_python/11400.py
"""
input :
7 8
1 4
4 5
5 1
1 6
6 7
2 7
7 3
2 3

output :
2
1 6
6 7
"""
import sys
sys.stdin = open("input.txt", "r")
input = sys.stdin.readline
sys.setrecursionlimit(10 ** 5)


def dfs(p, mp=-1):
    global NODE_NUM; global BCC_NUM;
    NODE_NUM += 1
    dfn[p] = low[p] = NODE_NUM
    for chd in graph[p]:
        if chd == mp: continue

        if not dfn[chd]:  # 처음 방문 Tree edge
            ST.append([p, chd])

            dfs(chd, p)
            low[p] = min(low[p], low[chd])

            # BCC 생성
            if dfn[p] <= low[chd]:
                # 여기서 단절선 감지!!
                if dfn[p] < low[chd]: CUT_EDGES.append([p, chd])

                BCC_NUM += 1
                while True:
                    [a, b] = ST.pop()
                    BCC[BCC_NUM].append([a, b])
                    if [a, b] == [p, chd]: break

        elif dfn[p] > dfn[chd]:
            # 간선의 중복방문을 막기위해 자식에서 조상으로 가는 경우만 관찰
            # 즉 여기선 chd가 chd가 아님
            low[p] = min(low[p], dfn[chd])
            ST.append([p, chd])


if __name__ == "__main__":
    V, E = map(int, input().rstrip().split())
    graph = dict()
    for i in range(1, V + 1):
        graph[i] = list()

    for _ in range(E):
        a, b = map(int, input().rstrip().split())
        graph[a].append(b)
        graph[b].append(a)

    BCC = [[] for _ in range(V + 1)]
    dfn = [0 for _ in range(V + 1)]
    low = [0 for _ in range(V + 1)]
    CUT_EDGES = []
    ST = []
    NODE_NUM = 0
    BCC_NUM = 0
    for i in range(1, V + 1):
        if dfn[i]: continue

        dfs(i, -1)

    # print ans
    print(len(CUT_EDGES))
    for edge in CUT_EDGES:
        edge.sort()
    CUT_EDGES.sort()
    for edge in CUT_EDGES:
        print(edge[0], edge[1])

File: 2021_summer_algorithm_study_intermediate/using_cpp/11400.cpp
/*
input :
7 8
1 4
4 5
5 1
1 6
6 7
2 7
7 3
2 3

output :
2
1 6
6 7
*/

#include <iostream>
#include <algorithm>
#include <vector>
#define SZ (100005)
using namespace std;

typedef struct edge{
    int a, b;
}edge;

// 전역 변수 선언
int V, E;
int NODE_NUM, BCC_NUM;
vector<int> graph[SZ];
vector<edge> BCC[SZ];
int dfn[SZ], low[SZ];
vector<edge> ST;
vector<edge> CUT_EDGES;

void dfs(int p, int mp){
    dfn[p] = low[p] = ++NODE_NUM;
    for (int chd: graph[p]){
        if (chd == mp) continue;

        if (dfn[chd] == 0){
            ST.push_back({p, chd});
            dfs(chd, p);
            low[p] = min(low[p], low[chd]);

            // BCC 감지
            if (dfn[p] <= low[chd]){
                if (dfn[p] < low[chd]) CUT_EDGES.push_back({p, chd});

                BCC_NUM++;
                while (true){
                    edge ab = ST.back(); ST.pop_back();
                    BCC[BCC_NUM].push_back(ab);

                    if (ab.a == p && ab.b == chd) break;
                }
            }
        }
        else if (dfn[p] > dfn[chd]){
            low[p] = min(low[p], dfn[chd]);
            ST.push_back({p, chd});
        }
    }
}


bool comp(edge obj1, edge obj2){
    if (obj1.a == obj2.a) return obj1.b < obj2.b;
    return obj1.a < obj2.a;
}

int main(){
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    freopen("input.txt", "r", stdin);

    cin >> V >> E;
    for (int i = 0; i < E; i++){
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    for (int i = 1; i <= V ; i++){
        if (dfn[i]) continue;

        dfs(i, -1);
    }

    // print ans
    cout << CUT_EDGES.size() << '\n';
    for (edge &ab: CUT_EDGES){
        if (ab.a > ab.b) {
            swap(ab.a, ab.b);
        }
    }
    sort(CUT_EDGES.begin(), CUT_EDGES.end(), comp);

    for (edge ab: CUT_EDGES){
        cout << ab.a << ' ' << ab.b << '\n';
    }

    return 0;
}